File: .claude/skills/query-opensearch/query.py
#!/usr/bin/env python3
"""Query the Open Pulse OpenSearch cluster.

Reads OPENSEARCH_ENDPOINT, OPENSEARCH_USERNAME, OPENSEARCH_PASSWORD
from .env at the repo root, or from the environment. The endpoint is
HTTPS with a self-signed cert by default, so TLS verification is
disabled (override with --verify).

Usage:
    # Cluster health
    python query.py health

    # List indices
    python query.py indices

    # Count docs in an index
    python query.py count <index>

    # DSL search (body from CLI, stdin, or file)
    python query.py search <index> '{"query":{"match_all":{}}}'
    python query.py search <index> -                 # stdin
    python query.py search <index> -f path.json      # file

    # Arbitrary GET — useful for _cat/* and one-off introspection
    python query.py get /<path>

Each subcommand prints JSON to stdout. Errors go to stderr with the
HTTP status and response body.
"""
from __future__ import annotations

import argparse
import base64
import json
import os
import ssl
import sys
import urllib.error
import urllib.parse
import urllib.request
from pathlib import Path


def load_dotenv() -> None:
    here = Path(__file__).resolve()
    for parent in [here.parent, *here.parents]:
        env = parent / ".env"
        if env.is_file():
            for raw in env.read_text().splitlines():
                line = raw.strip()
                if not line or line.startswith("#") or "=" not in line:
                    continue
                key, _, value = line.partition("=")
                os.environ.setdefault(key.strip(), value.strip())
            return


def request(method: str, path: str, body: bytes | None, verify: bool) -> dict:
    endpoint = os.environ["OPENSEARCH_ENDPOINT"].rstrip("/")
    user = os.environ["OPENSEARCH_USERNAME"]
    password = os.environ["OPENSEARCH_PASSWORD"]
    token = base64.b64encode(f"{user}:{password}".encode()).decode()
    headers = {"Authorization": f"Basic {token}", "Accept": "application/json"}

    proxy = os.environ.get("OPENSEARCH_DASHBOARDS_PROXY", "").strip().lower() in ("1", "true", "yes")
    if proxy:
        # OpenSearch Dashboards' Dev Tools proxy. The cluster API path goes
        # in the `path` query param (URL-encoded so its own `?query` survives),
        # the inner verb in `method`, and the outer request is always POST.
        inner = path[1:] if path.startswith("/") else path
        proxy_qs = urllib.parse.urlencode({"path": inner, "method": method.upper()})
        url = f"{endpoint}/api/console/proxy?{proxy_qs}"
        outer_method = "POST"
        headers["osd-xsrf"] = "true"
        # The proxy expects a content-type on POSTs even when the inner verb
        # is GET, so always set it.
        headers.setdefault("Content-Type", "application/json")
    else:
        url = f"{endpoint}{path}"
        outer_method = method
        if body is not None:
            headers["Content-Type"] = "application/json"

    req = urllib.request.Request(url, data=body, headers=headers, method=outer_method)
    ctx = ssl.create_default_context()
    if not verify:
        ctx.check_hostname = False
        ctx.verify_mode = ssl.CERT_NONE
    try:
        with urllib.request.urlopen(req, timeout=30, context=ctx) as resp:
            raw = resp.read()
    except urllib.error.HTTPError as e:
        print(f"http {e.code}: {e.read().decode(errors='replace')}", file=sys.stderr)
        sys.exit(1)
    except urllib.error.URLError as e:
        print(f"network error: {e.reason}", file=sys.stderr)
        sys.exit(1)

    if not raw:
        return {}
    try:
        return json.loads(raw)
    except json.JSONDecodeError:
        return {"_raw": raw.decode(errors="replace")}


def read_body(arg: str | None, file: str | None) -> bytes | None:
    if file:
        return Path(file).read_bytes()
    if arg == "-":
        return sys.stdin.buffer.read()
    if arg is None:
        return None
    return arg.encode()


def main() -> int:
    parser = argparse.ArgumentParser(description=__doc__.splitlines()[0])
    parser.add_argument("--verify", action="store_true", help="enforce TLS verification (default: skip)")
    sub = parser.add_subparsers(dest="cmd", required=True)

    sub.add_parser("health")
    p_indices = sub.add_parser("indices")
    p_indices.add_argument("pattern", nargs="?", default="*")

    p_count = sub.add_parser("count")
    p_count.add_argument("index")

    p_search = sub.add_parser("search")
    p_search.add_argument("index")
    p_search.add_argument("body", nargs="?", help="JSON body, or '-' for stdin")
    p_search.add_argument("-f", "--file")
    p_search.add_argument("--size", type=int, default=10)

    p_get = sub.add_parser("get")
    p_get.add_argument("path", help="path under the endpoint, eg /_cat/indices?v")

    args = parser.parse_args()
    load_dotenv()
    for key in ("OPENSEARCH_ENDPOINT", "OPENSEARCH_USERNAME", "OPENSEARCH_PASSWORD"):
        if not os.environ.get(key):
            print(f"error: {key} must be set", file=sys.stderr)
            return 2

    if args.cmd == "health":
        payload = request("GET", "/_cluster/health", None, args.verify)
    elif args.cmd == "indices":
        payload = request("GET", f"/_cat/indices/{args.pattern}?format=json&h=index,docs.count,store.size&s=index", None, args.verify)
    elif args.cmd == "count":
        payload = request("GET", f"/{args.index}/_count", None, args.verify)
    elif args.cmd == "search":
        body = read_body(args.body, args.file) or b'{"query":{"match_all":{}}}'
        payload = request("POST", f"/{args.index}/_search?size={args.size}", body, args.verify)
    elif args.cmd == "get":
        path = args.path if args.path.startswith("/") else f"/{args.path}"
        payload = request("GET", path, None, args.verify)
    else:
        parser.print_help()
        return 2

    json.dump(payload, sys.stdout, indent=2, default=str)
    sys.stdout.write("\n")
    return 0


if __name__ == "__main__":
    sys.exit(main())

File: .agents/skills/query-opensearch/query.mjs
#!/usr/bin/env node
// Query the Open Pulse OpenSearch cluster.
//
// Reads OPENSEARCH_ENDPOINT, OPENSEARCH_USERNAME, OPENSEARCH_PASSWORD
// from the nearest .env (walking up) or from process.env. TLS
// verification is disabled by default (self-signed); pass --verify to
// re-enable.
//
// Usage:
//   node query.mjs health
//   node query.mjs indices [pattern]
//   node query.mjs count <index>
//   node query.mjs search <index> '<json body>'
//   node query.mjs search <index> -                   # stdin
//   node query.mjs search <index> -f body.json
//   node query.mjs get /<path>

import { readFile, stat } from 'node:fs/promises';
import { dirname, join, resolve } from 'node:path';
import { fileURLToPath } from 'node:url';

async function loadDotenv() {
	let dir = dirname(fileURLToPath(import.meta.url));
	for (let i = 0; i < 10; i++) {
		const envPath = join(dir, '.env');
		try {
			await stat(envPath);
			const text = await readFile(envPath, 'utf8');
			for (const line of text.split('\n')) {
				const trimmed = line.trim();
				if (!trimmed || trimmed.startsWith('#') || !trimmed.includes('=')) continue;
				const idx = trimmed.indexOf('=');
				const key = trimmed.slice(0, idx).trim();
				const value = trimmed.slice(idx + 1).trim();
				if (process.env[key] === undefined) process.env[key] = value;
			}
			return;
		} catch {
			const parent = dirname(dir);
			if (parent === dir) return;
			dir = parent;
		}
	}
}

async function readStdin() {
	let data = '';
	for await (const chunk of process.stdin) data += chunk;
	return data;
}

async function request(method, path, body, { verify }) {
	const endpoint = process.env.OPENSEARCH_ENDPOINT.replace(/\/$/, '');
	const token = Buffer.from(`${process.env.OPENSEARCH_USERNAME}:${process.env.OPENSEARCH_PASSWORD}`).toString('base64');
	const headers = { Authorization: `Basic ${token}`, Accept: 'application/json' };

	const proxy = ['1', 'true', 'yes'].includes((process.env.OPENSEARCH_DASHBOARDS_PROXY || '').trim().toLowerCase());

	let url;
	let outerMethod;
	if (proxy) {
		// OpenSearch Dashboards' Dev Tools proxy: cluster path goes in
		// ?path=..., inner verb in ?method=..., outer request is always POST.
		const inner = path.startsWith('/') ? path.slice(1) : path;
		const qs = new URLSearchParams({ path: inner, method: method.toUpperCase() });
		url = `${endpoint}/api/console/proxy?${qs.toString()}`;
		outerMethod = 'POST';
		headers['osd-xsrf'] = 'true';
		// The proxy wants a content-type even when the inner verb is GET.
		if (!headers['Content-Type']) headers['Content-Type'] = 'application/json';
	} else {
		url = `${endpoint}${path}`;
		outerMethod = method;
		if (body !== undefined) headers['Content-Type'] = 'application/json';
	}

	// Disable TLS verification for self-signed certs by default.
	const previous = process.env.NODE_TLS_REJECT_UNAUTHORIZED;
	if (!verify) process.env.NODE_TLS_REJECT_UNAUTHORIZED = '0';

	let res;
	try {
		res = await fetch(url, { method: outerMethod, headers, body });
	} catch (e) {
		console.error(`network error: ${e.message}`);
		process.exit(1);
	} finally {
		if (previous === undefined) delete process.env.NODE_TLS_REJECT_UNAUTHORIZED;
		else process.env.NODE_TLS_REJECT_UNAUTHORIZED = previous;
	}

	if (!res.ok) {
		console.error(`http ${res.status}: ${await res.text()}`);
		process.exit(1);
	}
	const text = await res.text();
	try {
		return JSON.parse(text);
	} catch {
		return { _raw: text };
	}
}

async function readBody(arg, file) {
	if (file) return await readFile(resolve(file), 'utf8');
	if (arg === '-') return await readStdin();
	if (arg === undefined) return undefined;
	return arg;
}

function flag(argv, name) {
	const i = argv.indexOf(name);
	return i === -1 ? null : argv[i + 1];
}

async function main() {
	const argv = process.argv.slice(2);
	await loadDotenv();
	for (const key of ['OPENSEARCH_ENDPOINT', 'OPENSEARCH_USERNAME', 'OPENSEARCH_PASSWORD']) {
		if (!process.env[key]) {
			console.error(`error: ${key} must be set`);
			process.exit(2);
		}
	}
	const verify = argv.includes('--verify');
	const positional = argv.filter((a) => !a.startsWith('-') && a !== flag(argv, '-f'));
	const [cmd, ...rest] = positional;

	let payload;
	switch (cmd) {
		case 'health':
			payload = await request('GET', '/_cluster/health', undefined, { verify });
			break;
		case 'indices': {
			const pattern = rest[0] || '*';
			payload = await request(
				'GET',
				`/_cat/indices/${pattern}?format=json&h=index,docs.count,store.size&s=index`,
				undefined,
				{ verify }
			);
			break;
		}
		case 'count':
			if (!rest[0]) {
				console.error('usage: count <index>');
				process.exit(2);
			}
			payload = await request('GET', `/${rest[0]}/_count`, undefined, { verify });
			break;
		case 'search': {
			if (!rest[0]) {
				console.error('usage: search <index> [body|-|-f file]');
				process.exit(2);
			}
			const size = flag(argv, '--size') || '10';
			const body = (await readBody(rest[1], flag(argv, '-f'))) || '{"query":{"match_all":{}}}';
			payload = await request('POST', `/${rest[0]}/_search?size=${size}`, body, { verify });
			break;
		}
		case 'get':
			if (!rest[0]) {
				console.error('usage: get /<path>');
				process.exit(2);
			}
			payload = await request('GET', rest[0].startsWith('/') ? rest[0] : `/${rest[0]}`, undefined, { verify });
			break;
		default:
			console.error('usage: health | indices [pattern] | count <index> | search <index> [body] | get /<path>');
			process.exit(2);
	}

	process.stdout.write(JSON.stringify(payload, null, 2) + '\n');
}

main();
